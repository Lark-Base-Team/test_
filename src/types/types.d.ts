export interface SheetInfo {
  name: string;
  tableData: {
    fields: {
      name: string;
    }[];
    records: {
      [key: string]: string;
    }[];
  };
}

export interface ExcelDataInfo {
  sheets: SheetInfo[];
  name: string;
}

export interface IFieldValue {
  record_id: string;
  value: IOpenCellValue;
}

export interface IUndefinedFieldValue {
  record_id: undefined;
  value: undefined;
}

export interface fieldMap {
  field: IFieldMeta;
  excel_field: string;
  config: {
    format?: string;
    separator?: string;
    boolValue?: {
      true: string[];
      false: string[];
    }
    [key: string]: any;
  }
}