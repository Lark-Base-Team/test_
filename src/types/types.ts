import {
  IDuplexLinkFieldMeta,
  IFieldMeta,
  ISingleLinkFieldMeta,
  ITable,
  IField,
  IOpenCellValue,
} from "@lark-base-open/js-sdk"
import { TaskAction } from "@/utils/import/tasks"

export interface SheetInfo {
  name: string
  tableData: {
    fields: {
      name: string
    }[]
    records: {
      [key: string]: string
    }[]
  }
}

export interface ExcelDataInfo {
  sheets: SheetInfo[]
  name: string
}

export interface IFieldValue {
  record_id: string
  value: IOpenCellValue
}

export interface IUndefinedFieldValue {
  record_id: undefined
  value: undefined
}

export type LinkField = ISingleLinkFieldMeta | IDuplexLinkFieldMeta
export interface fieldMap {
  /**
   * Key for tree node
   *
   * @Description The unique key of the node in a tree table
   */
  key: string
  field: IFieldMeta
  table: string
  excel_field: string | undefined
  root: boolean
  config: {
    format?: string
    separator?: string
    boolValue?: {
      true: string[]
      false: string[]
    }
    primaryKey?: string
  }
  hasChildren: boolean
  children?: fieldMap[]
}

export interface BitableTable {
  table: ITable
  id: string
  name: string
  indexId: string[]
  root: boolean
  fields: {
    [key: string]: IField
  }
}

export interface Task<T = any> {
  table: {
    name: string
    id: string
  }
  action: TaskAction
  data: T[]
  result: boolean[] | undefined
}
