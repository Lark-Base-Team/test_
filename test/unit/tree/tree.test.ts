import { FieldType } from "@lark-base-open/js-sdk"
import { fieldMap } from "../../../src/types/types"
import {
  walkTree,
  getTreeLength,
  clearTree,
  getChild,
  getTreeChildren,
  getTreeDepth,
} from "../../../src/utils/import/tree"
import { describe, it, expect } from "vitest"

describe("tree test", () => {
  const fieldsMaps: fieldMap[] = [
    {
      key: "tbl",
      table: "tbl",
      hasChildren: true,
      root: true,
      excel_field: "tbl",
      field: {
        type: FieldType.SingleLink,
        name: "tbl",
        isPrimary: false,
        id: "field_id",
        description: {
          content: [],
        },
        property: {
          tableId: "tbl",
          multiple: false,
        },
      },
      config: {},
      children: [
        {
          key: "tbl1",
          table: "tbl2",
          root: false,
          hasChildren: false,
          excel_field: undefined,
          field: {
            name: "tbl1",
            type: FieldType.Text,
            id: "field_id",
            property: null,
            isPrimary: true,
            description: {
              content: [],
            },
          },
          config: {},
        },
        {
          key: "tbl2",
          table: "tbl3",
          root: false,
          hasChildren: true,
          excel_field: undefined,
          field: {
            name: "tbl1",
            type: FieldType.SingleLink,
            id: "field_id",
            property: {
              tableId: "tbl",
              multiple: false,
            },
            isPrimary: false,
            description: {
              content: [],
            },
          },
          config: {},
          children: [
            {
              key: "tbl3",
              table: "tbl1",
              root: true,
              hasChildren: false,
              excel_field: "tbl1",
              field: {
                name: "tbl1",
                type: FieldType.Text,
                id: "field_id",
                property: null,
                isPrimary: false,
                description: {
                  content: [],
                },
              },
              config: {},
            },
          ],
        },
      ],
    },
    {
      key: "tbl3",
      table: "tbl1",
      root: true,
      hasChildren: false,
      excel_field: "tbl1",
      field: {
        name: "tbl1",
        type: FieldType.Text,
        id: "field_id",
        property: null,
        isPrimary: false,
        description: {
          content: [],
        },
      },
      config: {},
    },
  ]

  it("clearTree", () => {
    const copy = JSON.parse(JSON.stringify(fieldsMaps))
    clearTree(copy)
    expect(copy).toMatchSnapshot()
    expect(
      Object.is(copy[0].excel_field, copy[0].children?.[0].excel_field),
    ).toBe(true)
  })

  it("walkTree: is same child", () => {
    const arr: any[] = []
    walkTree(fieldsMaps, (v: any) => {
      if (v.hasChildren) {
        arr.push(v)
      }
    })
    expect(Object.is(arr[0], fieldsMaps[0])).toBe(true)
  })

  it("getChild: is same child", () => {
    expect(
      Object.is(
        getChild(fieldsMaps, "tbl1"),
        (fieldsMaps[0].children as fieldMap["children"])?.[0],
      ),
    ).toBe(true)
  })

  it("getTreeChildren: is same child", () => {
    expect(
      Object.is(
        getTreeChildren(fieldsMaps, "tbl"),
        (fieldsMaps[0] as fieldMap).children,
      ),
    ).toBe(true)
  })

  it("getTreeLength", () => {
    expect(getTreeLength(fieldsMaps)).toBe(5)
  })

  it("getTreeDepth", () => {
    expect(getTreeDepth(fieldsMaps)).toBe(3)
  })
})
