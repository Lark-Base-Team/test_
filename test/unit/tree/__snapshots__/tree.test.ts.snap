// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tree test > clearTree 1`] = `
[
  {
    "children": [
      {
        "children": undefined,
        "config": {},
        "excel_field": "tbl",
        "field": {
          "description": {
            "content": [],
          },
          "id": "field_id",
          "isPrimary": true,
          "name": "tbl1",
          "property": null,
          "type": 1,
        },
        "hasChildren": false,
        "key": "tbl1",
        "root": false,
        "table": "tbl2",
      },
    ],
    "config": {},
    "excel_field": "tbl",
    "field": {
      "description": {
        "content": [],
      },
      "id": "field_id",
      "isPrimary": false,
      "name": "tbl",
      "property": {
        "multiple": false,
        "tableId": "tbl",
      },
      "type": 18,
    },
    "hasChildren": true,
    "key": "tbl",
    "root": true,
    "table": "tbl",
  },
  {
    "children": undefined,
    "config": {},
    "excel_field": "tbl1",
    "field": {
      "description": {
        "content": [],
      },
      "id": "field_id",
      "isPrimary": false,
      "name": "tbl1",
      "property": null,
      "type": 1,
    },
    "hasChildren": false,
    "key": "tbl3",
    "root": true,
    "table": "tbl1",
  },
]
`;
